OpenSim/Actuators/Millard2012EquilibriumMuscle.cpp

```diff
@@ -147,34 +151,44 @@ void Millard2012EquilibriumMuscle::extendFinalizeFromProperties()
     fpeCurve.ensureCurveUpToDate();
     fseCurve.ensureCurveUpToDate();
 
-    // Set properties of pennation model subcomponent. Values of
-    // optimal_fiber_length, pennation_angle_at_optimal, and
-    // maximum_pennation_angle are checked by
-    // MuscleFixedWidthPennationModel::extendFinalizeFromProperties().
+    // Propagate properties down to pennation model subcomponent. If any of the
+    // new property values are invalid, restore the subcomponent's current
+    // property values (to avoid throwing again when the subcomponent's
+    // extendFinalizeFromProperties() method is called directly) and then
+    // re-throw the exception thrown by the subcomponent.
     auto& penMdl =
         updMemberSubcomponent<MuscleFixedWidthPennationModel>(penMdlIdx);
+    MuscleFixedWidthPennationModel* penMdlCopy = penMdl.clone();
```

Comment:
This memory is leaked (ditto for `actMdlCopy`). Consider:

```cpp
std::unique_ptr<MuscleFixedWidthPennationModel> penMdlCopy(penMdl.clone());
```

(no need to change the copy assignment below; it will work as-is).

An alternative is to use the copy constructor:

```cpp
MuscleFixedWidthPennationModel penMdlCopy(penMdl);
```

Side note: The pattern of cloning and restoring is fine here because the specific subcomponent doesn't have sockets or add anything to the system; I'm not sure this pattern would work for a subcomponent with state variables.

Reply:
Good catch @chrisdembia. I think the pattern for any subcomponent would work, since it is a mechanism of caching and restoring its properties and not its state. I have slight preference for the copy constructor.

Reply:
Thanks @chrisdembia. I guess I misinterpreted Ajay's instructions. I'll change to copy constructor.

Reply:
@chrisdembia & @aseth1 - Have you ever considered removing `clone()` from the public interface? One could argue that I should have known better, but I wouldn't have eaten the poison if the API hadn't left it in the kitchen---and I don't use a poison detector in my workflow. User-centered design principle: error prevention.